### PR TITLE
Harden the agentic memory curator: allowlist, background pass, catch-up, dated sections

### PR DESCRIPTION
**Pull request by trieloff**
## Summary

Hardens the agentic memory curator against four failure modes observed in real runs: it was blocked on safe commands, never picked up sessions whose enrichment had failed, ran inline and blocked the cone, and worked against a "do not edit above this line" header that does not scale. It also fixes the delivery gap that would have kept all of that from reaching anyone who already has a workspace.

Before: a curator pass could stall the cone for minutes, silently skip every `pending-*.md` archive it had already failed once, and interrupt the user with approval prompts for `awk` and `echo`. After: it runs in the background, recovers pending archives on idle boot, maintains dated re-verifiable sections under a whole-file budget, and runs its analysis commands without escalating.

Cost measurement on a live instance then reshaped the last round of this PR. A single curator pass billed **$53.81** over 163 turns and 30 minutes, so the work turned to bringing the turn count down and to making sure nothing re-runs a pass automatically.

## Why

The four problems came out of a real on-call session (311 messages, $51.61) whose memory was never extracted. I then attached to a live instance over CDP to check the fixes against actual runs, which corrected three assumptions in the original draft:

1. **Three archives had been stuck for up to 17 days**, all flagged `pendingEnrichment`, including that on-call session. Nothing retried them. → `feat(webapp): retry pending session archives on idle boot`
2. **The curator rewrote the "protected" region anyway** despite the `preserve verbatim` instruction, and grew the file 31.8% in one pass. The live file measured 133.5% of its own budget with no enforcement, because `applyConeMemoryBudget` is only reachable from the legacy append paths and never runs after a curator pass. → dated sections + whole-file budget
3. **Two curator runs escalated six shell commands to the cone** (`awk` ×2, `echo` ×3, `sort` ×1), all read-only introspection, and the cone killed one run rather than grant shell access to an agent it did not recognise. → see below.
4. **`/shared/MEMORY.md` is seeded only when absent**, so every prompt change in this PR reached only brand-new workspaces. The live instance was still running the original prompt and the narrow 11-command allowlist. → `feat(shell): merge the curator contract into existing workspaces on upgrade`
5. **One curator pass cost $53.81** — 90% of what the entire 245-turn, 18-hour cone session that spawned it cost ($59.69), for one 30-minute sub-agent run. → `perf(scoops)` + `fix(webapp): never re-run the memory curator from the boot catch-up`

| | Curator, 1 pass | Cone, 245 turns / 18 h |
| --- | --- | --- |
| Cost | **$53.81** | $59.69 |
| Turns | 163 | 245 |
| Cache reads | 94,990,089 tok ($47.50, 88%) | 38,066,411 tok |
| Cache writes | 627,925 tok | 5,909,688 tok |

627,925 written against 94,990,089 read is one large context re-read ~151 times. **Turn count is the cost**, near-linearly.

## Approach / Implementation notes

**Why the allowlist has to be complete up front.** Non-cone scoops run with `defaultDisposition: 'require-approval'`, and `AlmostBashShell` deliberately skips its registration filter in that mode so an unlisted command escalates instead of dying as "command not found". For an unattended agent that inverts the intent: every gap becomes an approval prompt in the user's live conversation. Worse, the cone cannot grant its way out. It tried, reasoning *"I should approve these operations and set reads to always-allow so it doesn't keep prompting me"*, and the grant was persisted to `/scoops/agent-dapper-strawberry/etc/sudoers` — a folder named after a random per-run `agent-<adjective>-<flavor>` pair that is dropped when the run ends. The next run arrives under a name the cone has never seen. So the base set is code (`DEFAULT_ALLOWED_COMMANDS`), unioned with the user's frontmatter rather than replaced by it, so it reaches existing installs.

Rejected: widening the sandbox or giving the curator a non-escalating disposition. The `RestrictedFS` boundary is the real control, but changing the disposition would alter the security posture for every spawned agent, which is far more than this problem needs.

**Why the upgrade path, not an overwrite.** `MEMORY.md` is meant to be user-edited, so seeding over it would destroy local rules. `upgrade apply` already does a three-way merge against the bundled file at both release refs and classifies each path; adding `MEMORY.md` as a single-file scope alongside the three directory prefixes inherits `merged-clean` / `kept-local` / sidecar-on-conflict for free. This is a deliberate scope extension, and the skill doc is updated in both places that enumerate the scopes.

**Background pass.** The curator is spawned without awaiting it, so freeze no longer blocks on curation.

**Why the cost fixes live in the prompt.** Both turn-count drivers were instructions, not runtime:

- Spawned agents resolve an absent `thinkingLevel` to `'off'`, so the curator reasoned not at all and converged on the budget by trial and error, exactly as you saw: "cut 2000 more characters", then a turn that cut 200. It now defaults to `'medium'` (overridable per workspace) and is told to measure the surplus, budget the whole cut, then write once.
- The prompt said "read the archived session". Archives reach 2.5 MB, and the `slicc:session-data` block is a **single line** holding the whole session as JSON — about half the file. The new recipes strip that block and then drop `### Tool` sections, which are the bulk of the remaining bytes and hold no durable memory. Verified against three real archives: on the 2.5 MB one, orientation returns ~5 KB (0.22%) and the full conversation without tool output ~96 KB (3.9%). One session had 34 `## User` turns against 780 `### Tool` blocks.

**Why nothing retries a pass.** `timeoutSeconds` does not work and cannot be made to work from the caller. `waitForSpawn` resolves `{type: 'timeout'}` and abandons the promise; `AgentSpawnOptions` has no turn bound, deadline or `AbortSignal`, and `spawn()` returns only a promise, so there is no handle to cancel with. The measured run overran its 120s timeout by 14.9x. The boot catch-up therefore recovers pending archives only through the bounded single-call legacy enrichment, never by re-driving the agent. Real bounds on `agent` are filed as #1972.

Rejected: pinning the curator to a cheaper model. Cache reads dominate, but the fix for that is fewer turns over less context, and curation quality matters more than the per-token rate.

## Cross-cutting impact

- **Floats exercised**: logic is in `packages/webapp` (scoops + shell), so it is shared by CLI, extension, Electron and cloud alike. No float-specific wiring.
- **Other callers of the changed contract**: `upgrade apply`'s `SCOPES` is consumed only by `runtimePath()`, which returns `null` for out-of-scope paths; a new test pins that `/shared/CLAUDE.md` stays out of scope so the addition is not read as "all of `/shared/`".
- **Persisted state**: no schema change. `MEMORY.md` gains a frontmatter entry; the whole-file budget changes what the curator writes into `/workspace/CLAUDE.md`, not how it is stored.
- **Security**: the allowlist addition is `cp`, and the commands added are read-only text utilities. Nothing widens `writablePaths` (`/workspace/` only) or `visiblePaths` (`/sessions/`, `/shared/`). The self-protection on `/etc/sudoers*` is untouched.
- **Bundle size**: no new dependencies; all 6 budgets pass.

## Test plan

- [x] `npm run verify` — 7/7 checks pass (biome, prettier, custom-lints, boy-scout-debt, manifest, deadcode, deadcode-prod)
- [x] `node packages/dev-tools/tools/check-touched-exemptions.mjs` — OK, 0 changed files on any debt list
- [x] `npm run typecheck` — clean across all 9 projects
- [x] `npm run test` — 13,781 passing, no new failures
- [x] `npm run build`
- [x] `npm run build -w @slicc/chrome-extension`
- [x] `npm run bundle-size` — 6/6 budgets under limit
- [x] **New / changed behavior is covered by unit tests**: `npx vitest run packages/webapp/tests/scoops/agentic-memory.test.ts packages/webapp/tests/shell/supplemental-commands/upgrade-command.test.ts` — 29 passing. Covers each command granted from the base set when frontmatter omits it, the base set covering everything the seeded prompt is configured to use, the `MEMORY.md` three-way merge preserving local edits, and out-of-scope paths staying untouched.
- [x] Verified against a live instance over CDP: read the VFS archive index, the per-agent sudoers, and the cone's session history to confirm the diagnosis. I did **not** trigger a curator run — it costs real money and would rewrite the user's live memory file.
- [ ] N/A — no UI change, so no screencast.

**Pre-existing failures** (unrelated to this PR, now resolved upstream):

- `speech-e2e` was failing here and on every branch that triggered it since 2026-08-06. **Fixed on `main` by #1970**, now merged, and this branch is rebased onto it. Root cause was a race in `kokoro-js`'s vendored `phonemizer`: its emscripten build registers espeak's dictionaries asynchronously but publishes the instance without waiting, so a loaded-but-empty filesystem made `list_voices()` return `[]` and the one-shot language registry cached that emptiness for the page's lifetime. Load-sensitive, hence the unrelated branches. For the record, my earlier note on this PR blamed the `QuotaExceededError` in the log; #1970 disproved that with 4 GB of headroom.
- `packages/webapp/tests/speech/kokoro-voices.test.ts` fails locally only: `onnxruntime-node` ships no `darwin/arm64` `libonnxruntime.1.21.0.dylib` in this checkout. Environmental, passes in CI.
- Local coverage could not be run for the same reason, so `npm run test:coverage` is left to CI.

## Documentation

- [x] `docs/approvals.md` — documents that an "always" grant dies with an ephemeral agent's folder, so a recurring unattended agent must be configured up front rather than granted its way out
- [x] `packages/vfs-root/CLAUDE.md` — frontmatter `allowedCommands` extends the built-in base set rather than replacing it
- [x] `packages/vfs-root/workspace/skills/upgrade/SKILL.md` — `/shared/MEMORY.md` added to the documented scopes, in both the apply section and the "Do not" list
- [x] `packages/vfs-root/shared/MEMORY.md` — the curator contract itself

## Risk & rollback

- **Blast radius**: the curator runs on session freeze in every float. A bad prompt change shows up as a poor edit to `/workspace/CLAUDE.md`, not as data loss — the file is versioned in the VFS and the curator's own pass is what would be wrong.
- **Rollback**: revert cleanly. No persisted-state migration. The `upgrade` scope addition is inert until a user confirms an upgrade lick, and even then a conflict goes to a sidecar rather than the live file.
- **Worth a human check**: whether extending `upgrade apply` to a single file outside the three directory prefixes is the contract you want. The skill doc explicitly said not to modify files outside those prefixes "without the user explicitly extending the scope", and this PR is that extension.

## Checklist

- [x] Commits are focused and package-local where possible
- [x] No hand-edited files under `dist/`
- [x] No secrets, credentials, OAuth client secrets, or tokens in the diff
- [x] Public API / tool / shell-command / lick-channel changes are intentional and reflected in docs
- [x] If this changes a contract used by other floats, I checked the followers/leader/offscreen paths

---

### Review round 1 (Claude PR Review — LGTM, 3 🔵 minor)

| Finding | Resolution |
| --- | --- |
| Log wording went stale after the retry loop was added | Reworded to match post-retry behavior |
| `TODAY` is UTC-only, so section dates can be a day off locally | Documented as UTC in `packages/vfs-root/CLAUDE.md` |
| Flag-toggle rename edge case in the legacy catch-up path | Fixed to keep canonical archive names; regression test added |

### Review round 2 (live CDP diagnosis)

| Finding | Resolution |
| --- | --- |
| Curator escalated `awk` / `echo` / `sort` to the cone, which killed a run | Base allowlist already covers all three; added `cp` and documented why the list must lead the prompt |
| "Always" grants die with the agent's random per-run scoop folder | Documented in `docs/approvals.md`; the fix is a complete allowlist, not runtime grants |
| Prompt changes reached no existing workspace | `/shared/MEMORY.md` added to the `upgrade apply` scopes |

### Review round 3 (cost measurement on a live instance)

| Finding | Resolution |
| --- | --- |
| One pass billed $53.81; 88% of it cache reads, i.e. turn count | `thinkingLevel: medium` default so it plans instead of iterating |
| Prompt told it to read a 2.5 MB archive whose data block is one line | Verified reading recipes; orientation returns 0.22% of the file |
| `timeoutSeconds` is advisory — the run overran it 14.9x and kept billing | Boot catch-up no longer re-drives the curator; bounds filed as #1972 |

### Review round 4 (what memory is actually for)

The async curator told the cone nothing: `spawn` hardcoded `notifyOnComplete: false` and the pass discarded `finalText`, so a completed curation left only a log line.

| Finding | Resolution |
| --- | --- |
| Cone never learns a background pass ran | `notifyOnComplete` opt-in on `AgentSpawnOptions`, reusing the existing `scoop-notify` channel |
| Round 3's recipes dropped all `### Tool` blocks, discarding projects *and* pitfalls | Three targeted extractions instead, ~1% of the archive |
| Curator could install skills, since `/workspace/` was its writable root | Write grant narrowed to the memory file; `generateScoopSudoers` now emits the bare path so a single-file root works at all |

Memory now targets **preferences** (user turns), **projects** (top paths by frequency) and **pitfalls** (failed calls), and spends one `upskill search` on a pitfall when the session had failures, reporting any match in the message the cone now receives.

### CODE REVIEW

File: packages/webapp/src/ui/session-freezer.ts [hunk 1]
```diff
@@ -192,21 +202,30 @@ async function curateMemoriesBestEffort(
...
     });
-    return;
+    return null;
   }
   log.warn('Agentic memory pass failed after finishing — falling back to legacy extraction', {
     filename: frozen.filename,
     reason: result.reason,
   });
   await extractMemoriesBestEffort(opts, agentMessages, true);
+  return null;
```
[the remark this answers is not in the document]
Comment (trieloff) [reply packages/webapp/src/ui/session-freezer.ts:231]: Fixed in 83b41f3a7. The message now reads `Agentic memory pass unfinished — entry stays pending for boot catch-up` and carries `attemptLimit` so triage can see the retry bound.

The sibling case in the same function is fixed too: a pass that succeeded but whose index entry had already vanished was logged as a `warn` about a failed marker clear. The curated memory is on disk at that point, so it is now an `info` — `Agentic memory pass completed; index entry already gone`.

File: packages/webapp/src/scoops/agentic-memory.ts [hunk 2]
```diff
@@ -71,6 +82,7 @@ export async function runAgenticMemoryPass(
       SESSION_ARCHIVE_PATH: opts.sessionArchivePath,
       SESSION_COUNT: String(opts.sessionCount),
       BUDGET_CHARS: String(computeBudget(opts.sessionCount)),
+      TODAY: opts.today ?? new Date().toISOString().slice(0, 10),
```
[the remark this answers is not in the document]
Comment (trieloff) [reply packages/webapp/src/scoops/agentic-memory.ts:124]: Confirmed intentional, and now documented in 761723319. UTC is the right call here — it keeps the stamps comparable with archive `frozenAt` timestamps and makes the prompt deterministic under test via the `today` override.

Since users hand-edit these sections, the rule now says so explicitly in the curator prompt (`packages/vfs-root/shared/MEMORY.md`):

> Dates are UTC, matching the session archive timestamps, so a late-evening freeze west of UTC stamps the next day.

`packages/vfs-root/CLAUDE.md` carries the same note for developers.

File: packages/webapp/src/ui/session-freezer.ts [hunk 3]
```diff
@@ -745,16 +766,99 @@ async function updateSessionsIndex(
...
+        const entry = await recordPendingAttempt(opts.vfs, listedEntry.filename);
+        if (!entry) continue;
+        result.attempted += 1;
+        const updated = opts.agenticMemorySpawn
+          ? await runPendingAgenticMemoryPass(opts.vfs, entry, opts.agenticMemorySpawn)
+          : await enrichPendingSession(opts.vfs, entry, {
+              model: opts.model!,
+              apiKey: opts.apiKey!,
+              headers: opts.headers,
+            });
```
[the remark this answers is not in the document]
Comment (trieloff) [reply packages/webapp/src/ui/session-freezer.ts]: Good catch — this was the one real bug in the review. Fixed in 83b41f3a7, taking your option (a).

I keyed the guard on the filename rather than on which marker is set, because the filename is the direct evidence of whether a rename is legitimate:

```ts
const isDraft = isPendingDraftFilename(entry.filename);
const resolvedTitle = isDraft ? newTitle : entry.title;
const newFilename = isDraft
  ? `${entry.frozenAt.replace(/[:.]/g, '-')}-${slugify(newTitle)}.md`
  : entry.filename;
```

Only a `pending-…md` draft gets rewritten and renamed. A canonical archive still has its memory extracted and its markers cleared; it just keeps its name and title.

I went with (a) over (b) — gating the legacy fallback to `pendingEnrichment` entries only — because (b) would leave a `memoryPending`-only entry stuck pending forever if the flag never came back on, which is the exact problem this PR set out to fix. (a) still does the memory work, just without the harmful rename.

Worth noting your "most of the time the slug matches" read is why this needed an explicit test rather than an incidental one: the existing coverage passed either way. New regression test asserts the canonical name and title survive when the legacy title call returns something different:

`keeps a canonical curator archive at its name when legacy enrichment runs` in `packages/webapp/tests/ui/session-freezer.test.ts`.

**Comment by trieloff** [comment]
### Blocked by `speech-e2e`, not by anything in this PR

`ci` is red here solely because of `speech-e2e`. Every other job in run [31153985894](https://github.com/ai-ecoverse/slicc/actions/runs/31153985894) is green or skipped — 10 success, 16 skipped, 1 failure. Tracked by #1970.

**The failure is pre-existing.** `speech-e2e` is path-gated on webapp/speech changes, so most PRs skip it and `main` never goes red. On the branches that *do* trigger it, it has failed on every run since the evening of 2026-08-06, and passed on three consecutive runs earlier that same day. Nothing in this PR is implicated.

```
say: Invalid language identifier: "en-us". Should be one of: .
```

An empty language set from the phonemizer, after the model has already loaded. Deterministic — all three Playwright attempts, on two independent branches.

**Ruled out so far**, recorded so nobody repeats them:

| Suspect | Verdict |
| --- | --- |
| `vite` 8.1.5 → 8.2.0 (landed inside the regression window) | Ruled out — `kokoro-js`/`phonemizer` are installed at runtime by `ipk` from npm, not bundled by Vite, so it never touches espeak's data loading |
| A new npm release of an unpinned runtime dep | Ruled out — `kokoro-js`, `phonemizer`, `@huggingface/transformers`, `onnxruntime-web` all last published months before the window |
| Disk-derived browser storage quota | Ruled out by #1970's own instrumentation — freeing ~17 GB (`df` 83G → 100G avail) left usage at ~1 GB against a 4–5 GB quota with 3–4 GB headroom, and it still failed identically |

The `QuotaExceededError` console warning is real but is **not** global quota exhaustion, given that headroom.

**Unverified lead:** the logs also carry `An iframe which has both allow-scripts and allow-same-origin for its sandbox attribute can escape its sandboxing`. Sandboxed iframes can run on an opaque origin, where storage is restricted regardless of free disk — which would produce this exact shape. Stated as a hypothesis, not a finding.

No action needed on this branch. Once #1970 lands, re-run CI here and confirm the fix actually applied to *this* run rather than assuming it was inherited.
